worker/upgradedatabase/worker_test.go

```diff
@@ -268,25 +268,41 @@ func (s *workerSuite) TestNotPrimaryWatchForCompletionTimeout(c *gc.C) {
 	changes <- struct{}{}
 	s.watcher.EXPECT().Changes().Return(changes).AnyTimes()
 
+	timeout := make(chan time.Time, 1)
+	s.clock.EXPECT().After(gomock.Any()).Return(timeout)
+
 	// Primary does not complete the upgrade.
+	// After we have gone to the upgrade pending state, trip the timeout.
 	s.upgradeInfo.EXPECT().Refresh().Return(nil).AnyTimes()
-	s.upgradeInfo.EXPECT().Status().Return(state.UpgradePending).AnyTimes()
+	s.upgradeInfo.EXPECT().Status().DoAndReturn(func() state.UpgradeStatus {
+		// We only care about queueing one in the buffer.
+		// Carry on if we're jammed up - we'll fail elsewhere.
+		select {
+		case timeout <- time.Now():
+		default:
+		}
+		return state.UpgradePending
+	}).AnyTimes()
 
 	s.logger.EXPECT().Errorf("timed out waiting for primary database upgrade")
-	s.pool.EXPECT().SetStatus("0", status.Error, statusUpgrading)
+
+	finished := make(chan struct{})
+	s.pool.EXPECT().SetStatus("0", status.Error, statusUpgrading).Do(func(...interface{}) {
+		finished <- struct{}{}
```

Comment:
Did you intentionally block here (waiting for finished to be sent rather than just close(finished)) ?

I don't find the interactions for After to be particularly problematic. And I like that you can return a known channel that you can then trigger After() whenever you would like. (It could be in the main testing thread and not in a DoAndReturn() section, for example.

Reply:
I actually just replicated the coordination from a test below, knowing that we weren't blocking the worker loop there with the same technique. I think closing is a better idea though, so I'll do a follow-up.